src/server/player_sao.cpp

```diff
@@ -519,12 +519,15 @@ void PlayerSAO::rightClick(ServerActiveObject *clicker)
 
 void PlayerSAO::setHP(s32 target_hp, const PlayerHPChangeReason &reason, bool from_client)
 {
+	target_hp = rangelim(target_hp, -U16_MAX, U16_MAX);
+	s32 hp_change = target_hp - (s32)m_hp;
+
 	target_hp = rangelim(target_hp, 0, U16_MAX);
 
 	if (target_hp == m_hp)
 		return; // Nothing to do
 
-	s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, target_hp - (s32)m_hp, reason);
+	hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp_change, reason);
```

Comment:
The meaning of target_hp is different at different points of this code. It's first the unclamped HP value, then the clamped HP value. Should fix that to make the code simpler.